### PR TITLE
Move dispatcher into vortex io from vortex file

### PR DESCRIPTION
**Pull request by joseph-isaacs**

### CODE REVIEW

File: vortex-file/src/lib.rs [hunk 1]
```diff
@@ -196,9 +195,9 @@ mod forever_constant {
     }
 }
 
-pub use dispatcher::*;
 pub use forever_constant::*;
 pub use read::*;
+pub use vortex_io::dispatcher::*;
```
Comment (gatesn) [inline vortex-file/src/lib.rs]: I don't think we should re-export the dispatcher from this crate? Which I guess also means we shouldn't "forward" the features and instead the user should take a direct dependency on vortex-io and enable what they want.

File: vortex-io/src/lib.rs [hunk 2]
```diff
@@ -18,6 +18,7 @@ pub use write::*;
 mod buf;
 #[cfg(feature = "compio")]
 mod compio;
+pub mod dispatcher;
```
Comment (gatesn) [inline vortex-io/src/lib.rs]: `pub use dispatcher::*;`

**Comment by gatesn** [comment]
Best to wait for @a10y to comment I think

File: vortex-io/src/dispatcher/tokio.rs [hunk 3]
```diff
@@ -15,7 +15,7 @@ trait TokioSpawn {
 /// A [dispatcher][Dispatch] of IO operations that runs tasks on one of several
 /// Tokio `current_thread` runtimes.
 #[derive(Debug)]
-pub(super) struct TokioDispatcher {
+pub struct TokioDispatcher {
```
Comment (a10y) [inline vortex-io/src/dispatcher/tokio.rs]: The intention was that the different variants would stay private, and you would use the type erased `IoDispatcher` factory method to create the one you want. Is there a reason we need to expose this publicly?

Comment (joseph-isaacs) [reply vortex-io/src/dispatcher/tokio.rs]: Nope just a bad C&P

**Review by a10y — APPROVED** [review]

**Comment by a10y** [comment]
I forgot that we use IoDispatcher::new_tokio by default in the VortexReadBuilder and thus tokio can't be an optional dep, good catch

**Review by robert3005 — APPROVED** [review]

**Comment by robert3005** [comment]
@a10y we should have default for IoDispatcher which would pick based on features so you don't have to expose them directly to caller

**Comment by a10y** [comment]
Features are additive, so when you build and enable both features, which would it pick in that case?

**Comment by robert3005** [comment]
you're correct, no clue

**Pull request merged**